libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/ParametricSpec.hs

```diff
@@ -0,0 +1,301 @@
+--   The class EraOut (with method 'correctTxOut') supports specifcations over the type Family TxOut.
+--   Additional support for phased out Type Families like InstantaneousRewards,
+--   GenDelegs, FutureGenDelegs, StashedAVVMAddresses, and Ptrs, are handled by methods in EraOut
+module Test.Cardano.Ledger.Constrained.Conway.ParametricSpec (
+  module SimplePParams,
+  EraOut (..),
+  EraCert (..),
+  EraDeleg (..),
+  delegatedStakeReference,
+  CertKey (..),
+) where
+
+import Cardano.Ledger.Allegra (Allegra)
+import Cardano.Ledger.Alonzo (Alonzo)
+import Cardano.Ledger.Alonzo.TxOut (AlonzoEraTxOut (..), AlonzoTxOut (..))
+import Cardano.Ledger.Babbage (Babbage)
+import Cardano.Ledger.Babbage.TxOut (BabbageTxOut (..))
+import Cardano.Ledger.BaseTypes hiding (inject)
+import Cardano.Ledger.CertState
+import Cardano.Ledger.Coin (Coin (..), DeltaCoin (..))
+import Cardano.Ledger.Conway (Conway)
+import Cardano.Ledger.Conway.Rules (CertEnv (..))
+import Cardano.Ledger.Conway.TxCert
+import Cardano.Ledger.Core
+import Cardano.Ledger.Credential (Credential, StakeReference (..))
+import Cardano.Ledger.Crypto (Crypto)
+import Cardano.Ledger.Keys (KeyHash, KeyRole (..))
+import Cardano.Ledger.Mary (Mary, MaryValue)
+import Cardano.Ledger.PoolParams (PoolParams (ppId))
+import Cardano.Ledger.Shelley (Shelley)
+import Cardano.Ledger.Shelley.LedgerState (AccountState (..), StashedAVVMAddresses)
+import Cardano.Ledger.Shelley.TxCert
+import Cardano.Ledger.Shelley.TxOut (ShelleyTxOut (..))
+import Constrained hiding (Value)
+import Data.Map (Map)
+import qualified Data.Map as Map
+import Data.Word (Word64)
+import Test.Cardano.Ledger.Constrained.Conway.Cert (conwayTxCertSpec, shelleyTxCertSpec)
+import Test.Cardano.Ledger.Constrained.Conway.Deleg (EraDeleg (..))
+import Test.Cardano.Ledger.Constrained.Conway.Instances (ConwayFn, IsConwayUniv, toDelta_)
+import Test.Cardano.Ledger.Constrained.Conway.SimplePParams
+import qualified Test.Cardano.Ledger.Constrained.Conway.SimplePParams as SimplePParams
+
+-- ===========================================================
+
+-- | The class EraOut supports Era parametric Specifications that
+--   primarily navigate the differences in types parameterized type Family TxOut.
+--   Additional support for phased out Type Families like InstantaneousRewards,
+--   GenDelegs, FutureGenDelegs, StashedAVVMAddresses, and Ptrs, are also provided
+class
+  ( HasSpec fn (TxOut era)
+  , IsNormalType (TxOut era)
+  , HasSpec fn (StashedAVVMAddresses era)
+  , EraTxOut era
+  , IsConwayUniv fn
+  , EraPP era
+  , EraDeleg era
+  ) =>
+  EraOut era fn
+  where
+  irewardSpec :: Term fn AccountState -> Specification fn (InstantaneousRewards (EraCrypto era))
+  hasPtrs :: proxy era -> Term fn Bool
+  correctTxOut ::
+    Term fn (Map (Credential 'Staking (EraCrypto era)) (KeyHash 'StakePool (EraCrypto era))) ->
+    Term fn (TxOut era) ->
+    Pred fn
+
+betterTxOutShelley ::
```

Comment:
If this `TxOut` spec is relevant only in the `Shelley` era, then wouldn't it make sense to fix the era to `Shelley`?

Reply:
ok, I see now that you also use the same spec for Allegra

Reply:
These specs vary by what a Value is, and what a TxOut is. Allegra and Shelly have the same combination.